plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/datastore/driver/StorPoolPrimaryDataStoreDriver.java

```diff
@@ -328,66 +373,107 @@ private StorPoolSetVolumeEncryptionAnswer createEncryptedVolume(DataStore dataSt
-                } else {
-                    StorPoolResizeVolumeCommand resizeCmd = new StorPoolResizeVolumeCommand(vol.getPath(), new StorageFilerTO(pool), vol.getSize(), payload.newSize, payload.shrinkOk,
-                            payload.instanceName, payload.hosts == null ? false : true);
-                    answer = (ResizeVolumeAnswer) storageMgr.sendToPool(pool, payload.hosts, resizeCmd);
+        try {
+            SpConnectionDesc conn = StorPoolUtil.getSpConnection(data.getDataStore().getUuid(), data.getDataStore().getId(), storagePoolDetailsDao, primaryStoreDao);
 
-                    if (answer == null || !answer.getResult()) {
-                        err = answer != null ? answer.getDetails() : "return a null answer, resize failed for unknown reason";
-                    } else {
-                        path = StorPoolUtil.devPath(StorPoolUtil.getNameFromResponse(resp, false));
-
-                        vol.setSize(payload.newSize);
-                        vol.update();
-                        if (payload.newMaxIops != null) {
-                            VolumeVO volume = volumeDao.findById(vol.getId());
-                            volume.setMaxIops(payload.newMaxIops);
-                            volumeDao.update(volume.getId(), volume);
-                        }
+            err = updateStorPoolVolume(err, vol, payload, conn);
+            if (err == null && needResize) {
+                err = notifyQemuForTheNewSize(data, err, vol, payload);
+            }
 
-                        updateStoragePool(vol.getPoolId(), payload.newSize - oldSize);
-                    }
+            if (err != null) {
+                // try restoring volume to its initial size
+                SpApiResponse response = StorPoolUtil.volumeUpdate(name, oldSize, true, oldMaxIops, conn);
+                if (response.getError() != null) {
+                    logger.debug(String.format("Could not resize StorPool volume %s back to its original size. Error: %s", name, response.getError()));
                 }
-                if (err != null) {
-                    // try restoring volume to its initial size
-                    resp = StorPoolUtil.volumeUpdate(name, oldSize, true, oldMaxIops, conn);
-                    if (resp.getError() != null) {
-                        logger.debug(String.format("Could not resize StorPool volume %s back to its original size. Error: %s", name, resp.getError()));
-                    }
-                }
-            } catch (Exception e) {
-                logger.debug("sending resize command failed", e);
-                err = e.toString();
             }
+        } catch (Exception e) {
+            logger.debug("sending resize command failed", e);
+            err = e.toString();
+        }
+        return err;
+    }
+
+    private String notifyQemuForTheNewSize(DataObject data, String err, VolumeObject vol, ResizeVolumePayload payload)
```

Comment:
```suggestion
    private String notifyQemuForTheNewSize(DataObject data, VolumeObject vol, ResizeVolumePayload payload)
```

seems `err` not needed in the call